### PR TITLE
Synapse fix

### PR DESCRIPTION
**Pull request by isabella618033**
- saving previous synapse request and reuse it for the same input 
- debugged so the time it takes to complete 2 synapse come down from 8-12 sec to 2-3 sec

### CODE REVIEW

File: bittensor/_neuron/text/template_server/nucleus_impl.py [hunk 1]
```diff
@@ -131,34 +132,39 @@ def encode_forward(self,inputs,tokenizer=None):
...
-        elif self.config.neuron.autocast and self.device == 'cuda':
-            pre_hidden = self.pre_model(inputs).last_hidden_state
+        if self.config.neuron.training or self.config.neuron.autocast and self.device == 'cuda':
+            output = self.pre_model(inputs, output_hidden_states=True)
+            pre_hidden = output.hidden_states[-1]
         else:
             with torch.no_grad():
-                pre_hidden = self.pre_model(inputs).last_hidden_state
-        
+                output = self.pre_model(inputs, output_hidden_states=True)
```
Comment (Eugene-hu) [inline bittensor/_neuron/text/template_server/nucleus_impl.py:146]: Would this always compute the logits layer as well? If so, do we know how much time the last layer takes?

File: bittensor/_neuron/text/template_server/nucleus_impl.py [hunk 2]
```diff
@@ -131,34 +132,39 @@ def encode_forward(self,inputs,tokenizer=None):
...
+                pre_hidden = output.hidden_states[-1]
+
         if self.interpolate and sen_len[1] != pre_hidden.size()[1]:
             down= F.interpolate(pre_hidden.unsqueeze(1),size=[sen_len[1],pre_hidden.size()[2]],mode=self.inter_degree).squeeze(1)
```
Comment (Eugene-hu) [inline bittensor/_neuron/text/template_server/nucleus_impl.py:150]: We can probably remove the padding, and interpolation here and move them to encode and decode functions in the synapses

File: bittensor/_neuron/text/template_server/run.py [hunk 3]
```diff
@@ -84,17 +84,22 @@ def forward_generate( inputs_x:torch.FloatTensor, synapse):
...
-
-    def forward_casual_lm(inputs_x, synapse):
-        output = model.pre_model(input_ids=inputs_x).logits
-        return output
+    def forward_hidden_state(inputs_x, synapse, model_output = None):
+        if model_output == None:
+            model_output = model.encode_forward(inputs_x.to(model.device))
+        
+        hidden = model_output.hidden
+        padding_r = (1024-hidden.size(2))
```
Comment (Eugene-hu) [inline bittensor/_neuron/text/template_server/run.py]: If we use the encode forward function, we can remove these padding functions.

**Review by Eugene-hu — APPROVED** [review]
Looks good to me! Left you a few comments

File: bittensor/_neuron/text/template_server/nucleus_impl.py (hunk 1 above)
Comment (isabella618033) [reply bittensor/_neuron/text/template_server/nucleus_impl.py:146]: pre_model() always compute the whole model, always including the last layer, setting option output_hidden_states = True just add a return variable.

File: bittensor/_neuron/text/template_server/run.py (hunk 3 above)
Comment (isabella618033) [reply bittensor/_neuron/text/template_server/run.py]: okie
